crates/pubsub/src/sub.rs

```diff
@@ -0,0 +1,236 @@
+/// local ID.
+///
+/// This type is mostly a wrapper around [`broadcast::Receiver`], and exposes
+/// the same methods.
+#[derive(Debug)]
+pub struct RawSubscription {
+    /// The channel via which notifications are received.
+    pub(crate) rx: broadcast::Receiver<Box<RawValue>>,
+    /// The local ID of the subscription.
+    pub(crate) local_id: B256,
+}
+
+impl RawSubscription {
+    /// Get the local ID of the subscription.
+    pub const fn local_id(&self) -> B256 {
+        self.local_id
+    }
+
+    /// Wrapper for [`blocking_recv`]. Block the current thread until a message
+    /// is available.
+    ///
+    /// [`blocking_recv`]: broadcast::Receiver::blocking_recv
+    pub fn blocking_recv(&mut self) -> Result<Box<RawValue>, broadcast::error::RecvError> {
+        self.rx.blocking_recv()
+    }
+
+    /// Returns `true` if the broadcast channel is empty (i.e. there are
+    /// currently no notifications to receive).
+    pub fn is_empty(&self) -> bool {
+        self.rx.is_empty()
+    }
+
+    /// Returns the number of messages in the broadcast channel that this
+    /// receiver has yet to receive.
+    pub fn len(&self) -> usize {
+        self.rx.len()
+    }
+
+    /// Wrapper for [`recv`]. Await an item from the channel.
+    ///
+    /// [`recv`]: broadcast::Receiver::recv
+    pub async fn recv(&mut self) -> Result<Box<RawValue>, broadcast::error::RecvError> {
+        self.rx.recv().await
+    }
+
+    /// Wrapper for [`resubscribe`]. Create a new Subscription, starting from
+    /// the current tail element.
+    ///
+    /// [`resubscribe`]: broadcast::Receiver::resubscribe
+    pub fn resubscribe(&self) -> Self {
+        Self { rx: self.rx.resubscribe(), local_id: self.local_id }
+    }
+
+    /// Wrapper for [`same_channel`]. Returns `true` if the two subscriptions
+    /// share the same broadcast channel.
+    ///
+    /// [`same_channel`]: broadcast::Receiver::same_channel
+    pub fn same_channel(&self, other: &Self) -> bool {
+        self.rx.same_channel(&other.rx)
+    }
+
+    /// Wrapper for [`try_recv`]. Attempt to receive a message from the channel
+    /// without awaiting.
+    ///
+    /// [`try_recv`]: broadcast::Receiver::try_recv
+    pub fn try_recv(&mut self) -> Result<Box<RawValue>, broadcast::error::TryRecvError> {
+        self.rx.try_recv()
+    }
+}
+
+#[derive(Debug)]
+/// An item in a typed [`Subscription`]. This is either the expected type, or
+/// some serialized value of another type.
+pub enum SubscriptionItem<T> {
+    /// The expected item.
+    Item(T),
+    /// Some other value.
+    Other(Box<RawValue>),
+}
+
+impl<T: DeserializeOwned> From<Box<RawValue>> for SubscriptionItem<T> {
+    fn from(value: Box<RawValue>) -> Self {
+        if let Ok(item) = serde_json::from_str(value.get()) {
+            SubscriptionItem::Item(item)
+        } else {
+            SubscriptionItem::Other(value)
+        }
+    }
+}
```

Comment:
this will silently drop deserialization errors in the Subscription

can we add `recv_item` to RawSubscription ?

Reply:
> this will silently drop deserialization errors in the Subscription

this is intentional, as a deser error implies the subscription is of an unexpected type

Reply:
> can we add recv_item to RawSubscription ?

what's the expected behavior?  something like `async fn recv_item<T>(&mut self) -> Result<SubscriptionItem<T>, RecvError>` ?

Reply:
yeah,

sometimes the json types are not quite right, so it'd be beneficial to also receive failed deserde items.

but technically this would then behave like a result, so maybe we can add a subscription type that returns results
not sure what the best ux here is

but for debugging, I think I'd want to quickly swap out the `subscribe` call to a different function that returns `Result<T, >` or subscription

Reply:
how about this api

`recv` - get next item, returning `T`, discarding all errors
`recv_result` - get next item, returning `Result<T. serde_json::Error>`
`recv_any` - get next item, returning `SubscriptionItem<T>`